src/rebar_plugins.erl

```diff
@@ -32,31 +32,49 @@ handle_plugins(Plugins, State) ->
 
 handle_plugins(Profile, Plugins, State) ->
     %% Set deps dir to plugins dir so apps are installed there
+    Locks = rebar_state:lock(State),
+    DepsDir = rebar_state:get(State, deps_dir, ?DEFAULT_DEPS_DIR),
     State1 = rebar_state:set(State, deps_dir, ?DEFAULT_PLUGINS_DIR),
 
-    PluginProviders = lists:flatmap(fun(Plugin) ->
-                                            handle_plugin(Profile, Plugin, State1)
-                                    end, Plugins),
+    {PluginProviders, State2} =
+        lists:foldl(fun(Plugin, {PluginAcc, StateAcc}) ->
+                            {NewPlugins, NewState} = handle_plugin(Profile, Plugin, StateAcc),
+                            {PluginAcc++NewPlugins, NewState}
+                      end, {[], State1}, Plugins),
 
     %% reset deps dir
-    State2 = rebar_state:set(State1, deps_dir, ?DEFAULT_DEPS_DIR),
+    State3 = rebar_state:set(State2, deps_dir, DepsDir),
+    State4 = rebar_state:lock(State3, Locks),
 
-    rebar_state:create_logic_providers(PluginProviders, State2).
+    rebar_state:create_logic_providers(PluginProviders, State4).
 
 handle_plugin(Profile, Plugin, State) ->
     try
-        {ok, _, State2} = rebar_prv_install_deps:handle_deps(Profile, State, [Plugin]),
+        {ok, Apps, State2} = rebar_prv_install_deps:handle_deps(Profile, State, [Plugin]),
+        {no_cycle, Sorted} = rebar_prv_install_deps:find_cycles(Apps),
+        ToBuild = rebar_prv_install_deps:cull_compile(Sorted, []),
```

Comment:
Would it be more efficient to run compile/cycle detection only after having fetched all plugin code? We could differentiate between a plugin not being available and a plugin dep cycle, and save on operations otherwise.

Or maybe I'm just missing something here.